### PR TITLE
Ticket1828 neocera ioc

### PR DESCRIPTION
**Pull request by John-Holt-Tessella**
### Description of work

Implement Neocera IOC, emulator and OPI

### Associated Work

https://github.com/ISISComputingGroup/EPICS-DbChecker/pull/2
https://github.com/ISISComputingGroup/EPICS-neocera (master but needs referencing from EPICS)
https://github.com/ISISComputingGroup/ibex_gui/pull/427
https://github.com/ISISComputingGroup/EPICS/pull/63
https://github.com/ISISComputingGroup/EPICS-DeviceEmulator/pull/8

### To test

https://github.com/ISISComputingGroup/IBEX/issues/1828

### Acceptance criteria

Can monitor and set values on the Neocera LTC21 as per lab view. See https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Neocera-LTC-21 for commands.

---

#### Code Review

- [x] Is the code of an acceptable quality?
- [x] Are the PVs named according to the [naming standards](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/PV-Naming).
- [x] Have suitable [disable records](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Disable-records) been added?
- [x] Have suitable [simulation records](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Record-Simulation) been added?
- [x] Have the [finishing touches](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/IOC-Finishing-Touches) been applied?
- [x] Is there an [emulator](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Emulating-Devices) for the device?
- [x] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?
- [x] If an OPI has been modified, does it conform to the [style guidelines](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation)? There is a script called `check_opi_format.py` to help with this.

### Functional Tests

- [x] Do changes function as described? Add comments below that describe the tests performed.
- [ ] Does the IOC respond correctly both in full and simulation mode, where it's possible to test both?
- [x] If there are multiple _0n IOCs, do they run correctly?

### Final steps

- [ ] Reviewer has updated the submodule in the main EPICS repo? See **Reviewing work for the subModules of EPICS** in the [Git workflow](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Git-workflow) page for details.
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

**Comment by DominicOram** [comment]
Small issues found so far:
- [ ] Setting a heater limit <0 or >100 does not change readback on device but gives no alarm
- [ ] Setting negative K temperatures causes the device to beep in error but does not cause alarm for user
- [ ] What is the activity PV for? I don't think we're using activity lights under IBEX as opposed to SECI

**Comment by John-Holt-Tessella** [comment]
Heater limit gets set to max or min if you go over it (default epics behaviour)
Negative K is hard there is an error lag but LabView doesn't use it so I haven't either I am inclined not to do this.
Activity PV removed
PID range is dyamic so not included

File: NEOCERA/iocBoot/iocNEOCERA-IOC-01/config.xml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" ?>
+<ioc_config xmlns:xi="http://www.w3.org/2001/XInclude">
+ <config_part>
+  <ioc_desc>Neocera LTC21</ioc_desc>
+  <ioc_details>https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Neocera-LTC-21</ioc_details>
+  <macros>
+    <macro name="PORT" pattern="^COM[0-9]+$" description="Com port for device." />
+    <macro name="BAUD" pattern="^[A-Z0-9]+$" description="Baud rate for device (default 9600)" />
```
Comment (AdrianPotter) [inline NEOCERA/iocBoot/iocNEOCERA-IOC-01/config.xml]: Missing comma after `default`

**Comment by AdrianPotter** [comment]
Sensor 2 is not showing any units. 

image

**Comment by AdrianPotter** [comment]
- [ ] Recsim control behaviour is inconsistent. Some SPs feed their respective RBVs but the following don't:
    - Control
    - I
    - D
    - Fixed power reset
    - Heater limit
    - Gain
    - Offset

File: NEOCERA/iocBoot/iocNEOCERA-IOC-01/st-common.cmd [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+$(IFNOTRECSIM) $(IFDEVSIM)  drvAsynIPPortConfigure("L0", "localhost:57677")
+
+< $(IOCSTARTUP)/dbload.cmd
+
+##
+##
+##
+##
+##
```
Comment (AdrianPotter) [inline NEOCERA/iocBoot/iocNEOCERA-IOC-01/st-common.cmd]: Are these necessary?

File: NEOCERA/iocBoot/iocNEOCERA-IOC-01/st-common.cmd [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+< $(IOCSTARTUP)/init.cmd
+
+$(IFNOTRECSIM) $(IFNOTDEVSIM) drvAsynSerialPortConfigure("L0", "$(PORT)", 0, 0, 0, 0)
+$(IFNOTRECSIM) $(IFNOTDEVSIM) asynSetOption("L0", -1, "baud", "$(BAUD=9600)")
+$(IFNOTRECSIM) $(IFNOTDEVSIM) asynSetOption("L0", -1, "bits", "8")
+$(IFNOTRECSIM) $(IFNOTDEVSIM) asynSetOption("L0", -1, "parity", "none")
+$(IFNOTRECSIM) $(IFNOTDEVSIM) asynSetOption("L0", -1, "stop", "1")
+
+$(IFNOTRECSIM) $(IFDEVSIM)  drvAsynIPPortConfigure("L0", "localhost:57677")
```
Comment (AdrianPotter) [inline NEOCERA/iocBoot/iocNEOCERA-IOC-01/st-common.cmd]: Use freeport rather than hard coded port

**Comment by AdrianPotter** [comment]
- [ ] Why is there no monitor/control button on the details dialog? It is the only control that doesn't appear there.

**Comment by AdrianPotter** [comment]
- [ ] Why is there no monitor/control toggle on the details page? Every other control is present there.

File: NEOCERA/iocBoot/iocNEOCERA-IOC-01/st-common.cmd (hunk 2 above)
Comment (John-Holt-Tessella) [inline NEOCERA/iocBoot/iocNEOCERA-IOC-01/st-common.cmd]: Yes they give a nice indicator of where the db load starts

File: NEOCERA/iocBoot/iocNEOCERA-IOC-01/st-common.cmd (hunk 3 above)
Comment (John-Holt-Tessella) [inline NEOCERA/iocBoot/iocNEOCERA-IOC-01/st-common.cmd]: This is going to be a role out when we have decided how to use that.

**Pull request merged**